File: labs/lab11/Valid_Parenthesis_String.py
class Solution(object):
    def checkValidString(self, s):
        """
        :type s: str
        :rtype: bool
        """
        low = 0  
        high = 0  

        for char in s:
            if char == '(':
                low += 1
                high += 1
            elif char == ')':
                low = max(low - 1, 0)
                high -= 1
            elif char == '*':
                low = max(low - 1, 0)  
                high += 1 

            if high < 0:
                return False

        return low == 0

File: labs/lab11/Valid_Parenthesis_String.js
var checkValidString = function (s) {
  let low = 0;
  let high = 0;

  for (const char of s) {
    if (char === "(") {
      low++;
      high++;
    } else if (char === ")") {
      low = Math.max(low - 1, 0);
      high--;
    } else if (char === "*") {
      low = Math.max(low - 1, 0);
      high++;
    }

    if (high < 0) {
      return false;
    }
  }

  return low === 0;
};

console.log(checkValidString("()")); // Output: true
console.log(checkValidString("(*)")); // Output: true
console.log(checkValidString("(*))")); // Output: true
